src/rp2_common/hardware_adc/include/hardware/adc.h

```diff
@@ -128,15 +128,24 @@ static inline void adc_set_temp_sensor_enabled(bool enable) {
  *
  *  Performs an ADC conversion, waits for the result, and then returns it.
  *
- * \return Result of the conversion.
+ * \return Result of the conversion or negative value on error.
  */
-static inline uint16_t adc_read(void) {
+static inline int32_t adc_read(void) {
     hw_set_bits(&adc_hw->cs, ADC_CS_START_ONCE_BITS);
+    
+    while ((adc_hw->cs & (ADC_CS_READY_BITS|ADC_CS_ERR_BITS|ADC_CS_ERR_STICKY_BITS))==0) {
+         tight_loop_contents();
+    }
 
-    while (!(adc_hw->cs & ADC_CS_READY_BITS))
-        tight_loop_contents();
-
-    return (uint16_t) adc_hw->result;
+    if ((adc_hw->cs & ADC_CS_READY_BITS)>0) {
+        return adc_hw->result & 0xfff;
```

Comment:
Might be better to use ADC_RESULT_BITS

Reply:
I will have a look if that will have benefit. The original code check for the ADC_CS_READY_BITS.

Reply:
I believe Peter meant use `ADC_RESULT_BITS` instead of `0xfff` - see https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2040/hardware_regs/include/hardware/regs/adc.h#L111

Reply:
Oh,  :)
Thanks for the eye opener.